File: Batch4ProblemSolving/string-compression.java
//443. String Compression
//https://leetcode.com/problems/string-compression
// ref: inspired by leetcode solutions 

class Solution {
    public int compress(char[] chars) {
        //edge cases
        if(chars.length == 1) {
            return 1;
        }
        int len = chars.length;
        int index = 0;
        int j = 0;
        
        while(j < len) {
            char lastChar = chars[j];
            int count = 0;
            
            while(j < len && chars[j] == lastChar) {
                j++; 
                count++;
            }
            
            chars[index++] = lastChar;
        
            if(count > 1) {
                for(char ch: String.valueOf(count).toCharArray()) {
                    chars[index++] = ch;
                }
            }
        }
        
        return index;
    }
}

File: Batch4ProblemSolving/string-compression.kt
//https://leetcode.com/problems/string-compression
class Solution {
    fun compress(chars: CharArray): Int {
        //step: handle edge cases 
        if(chars.size == 1) {
            return 1
        }
        
        var index = 0
        var i = 0
        while(i <= chars.lastIndex) {
            var count = 0
            var lastChar = chars[i]
            
            //iterate over until mismatch with last char 
            while(i <= chars.lastIndex && chars[i] == lastChar) {
                i++
                count++
            }
    
            chars[index++] = lastChar
            
            if(count > 1) {
                for(ch in count.toString().toCharArray()) {
                    chars[index++] = ch
                }
            }
        }
        
        return index
    }

}
